extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+      }
+
+      // DimensionRangeShardSpec only handles columns that appear as-is in the output.
+      if (outputColumns.isEmpty()) {
+        return Collections.emptyList();
+      }
+
+      shardColumns.add(outputColumns.get(0));
+    }
+
+    return shardColumns;
+  }
+
+  private static boolean isClusterByBoosted(final ClusterBy clusterBy)
+  {
+    return !clusterBy.getColumns().isEmpty()
+           && clusterBy.getColumns()
+                       .get(clusterBy.getColumns().size() - 1)
+                       .columnName()
+                       .equals(QueryKitUtils.PARTITION_BOOST_COLUMN);
+  }
+
+  private static StringTuple makeStringTuple(
+      final ClusterBy clusterBy,
+      final RowKeyReader keyReader,
+      final RowKey key
+  )
+  {
+    final String[] array = new String[clusterBy.getColumns().size() - clusterBy.getBucketByCount()];
+    final boolean boosted = isClusterByBoosted(clusterBy);
+
+    for (int i = 0; i < array.length; i++) {
+      final Object val = keyReader.read(key, clusterBy.getBucketByCount() + i);
+
+      if (i == array.length - 1 && boosted) {
+        // Boost column
+        //noinspection RedundantCast: false alarm; the cast is necessary
+        array[i] = StringUtils.format("%016d", (long) val);
+      } else {
+        array[i] = (String) val;
+      }
+    }
+
+    return new StringTuple(array);
+  }
+
+  private static Pair<List<DimensionSchema>, List<AggregatorFactory>> makeDimensionsAndAggregatorsForIngestion(
+      final RowSignature querySignature,
+      final ClusterBy queryClusterBy,
+      final List<String> segmentSortOrder,
+      final ColumnMappings columnMappings,
+      final boolean isRollupQuery,
+      final Query<?> query
+  )
+  {
+    final List<DimensionSchema> dimensions = new ArrayList<>();
+    final List<AggregatorFactory> aggregators = new ArrayList<>();
+
+    // During ingestion, segment sort order is determined by the order of fields in the DimensionsSchema. We want
+    // this to match user intent as dictated by the declared segment sort order and CLUSTERED BY, so add things in
+    // that order.
+
+    // Start with segmentSortOrder.
+    final Set<String> outputColumnsInOrder = new LinkedHashSet<>(segmentSortOrder);
+
+    // Then the query-level CLUSTERED BY.
+    // Note: this doesn't work when CLUSTERED BY specifies an expression that is not being selected.
+    for (final SortColumn clusterByColumn : queryClusterBy.getColumns()) {
+      if (clusterByColumn.descending()) {
+        throw new MSQException(new InsertCannotOrderByDescendingFault(clusterByColumn.columnName()));
+      }
+
+      outputColumnsInOrder.addAll(columnMappings.getOutputColumnsForQueryColumn(clusterByColumn.columnName()));
+    }
+
+    // Then all other columns.
+    outputColumnsInOrder.addAll(columnMappings.getOutputColumnNames());
+
+    Map<String, AggregatorFactory> outputColumnAggregatorFactories = new HashMap<>();
+
+    if (isRollupQuery) {
```

Comment:
Comment to say why rollup is special here?

Reply:
Added:

> Populate aggregators from the native query when doing an ingest in rollup mode.